http/server.lua

```diff
@@ -753,7 +754,43 @@ local function parse_request(req)
     return p
 end
 
+local function add_connection(self, fiber_obj)
+    local fiber_id = fiber_obj:id()
+    self.connections[fiber_id] = {
+        obj = fiber_obj,
+        last_time_active = fiber.clock(),
+    }
+
+    return true
+end
+
+local function close_connection(self, fiber_obj, fiber_id, logger)
+    -- Name of fiber for http connection contains a hostname and port of http client.
+    local fiber_name = fiber_obj:name()
+    self.connections[fiber_id] = nil
+    logger("Close connection with %s, idle_timeout was exceeded\n", fiber_name)
+    fiber_obj:cancel()
+
+    return true
+end
+
+local function check_connections(self)
+    local logger = get_request_logger(self.options)
+
+    -- FIXME: complexity is O(N) and it is bad
+    for fiber_id, opts in pairs(self.connections) do
+        if fiber.clock() - opts.last_time_active >= self.options.idle_timeout then
+            self:close_connection(opts.obj, fiber_id, logger)
+        end
+    end
+end
+
 local function process_client(self, s, peer)
+    if self.options.idle_timeout > 0 then
+        self:check_connections()
+        self:add_connection(fiber.self())
+    end
```

Comment:
This way we'll cancel long requests (when a client's code doing a long work) and slow requests (say, when writting a large response requires considerable time). Not only idle connections.

It seems, we can just set the `idle_timeout` as timeout for the `:read()` call several lines below and handle the timeout error properly.

Reply:
Patch updated (not squashed yet).

Reply:
> This way we'll cancel long requests (when a client's code doing a long work) and slow requests (say, when writting a large response requires considerable time). Not only idle connections.

Right, implementation was not ideal. It was an attempt to create a scheduler for cooperative multitasking.

Patch was rewritten and now uses `timeout` option in `tcp_server`.

Reply:
> Now you set the timeout for getaddinfo() at socket binding. It is completely unrelated here.

Right. https://github.com/tarantool/doc/issues/2946

Reply:
> The 'test_idle_timeout' test passes, because it does not test anything. The keepalive_* options are not set for the http client.

It is still so.

```diff
diff --git a/http/server.lua b/http/server.lua
index 08d59b5..3f8c887 100644
--- a/http/server.lua
+++ b/http/server.lua
@@ -762,7 +762,7 @@ local function process_client(self, s, peer)
                 delimiter = { "\n\n", "\r\n\r\n" },
                 -- Maximum number of seconds to wait,
                 -- for example 50 means "stop after 50 seconds".
-                timeout = self.idle_timeout,
+                -- timeout = self.idle_timeout,
             }
 
             if chunk == '' then
```

```shell
$ cmake . && make && luatest -v
<...>
Ran 54 tests in 0.543 seconds, 54 succeeded, 0 failed
```